### PR TITLE
Memory game

### PR DESCRIPTION
**Pull request by ustymchyk**
_Complete the form below and remove this line._

### Memory – Pair Game

[Demo](https://ustymchyk.github.io/memory-game/) |
[Code base](https://github.com/ustymchyk/memory-game/tree/master)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/ustymchyk/memory-pair-game/index.html [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <link rel="stylesheet" href="index.css">
+  <link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
+  <title>Memory game</title>
+</head>
+
+<body>
+  <ul class="game"></ul>
+</body>
+<script src="index.js"></script>
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.html:15]: Move this `script` tag before the `body` tag closing

File: submissions/ustymchyk/memory-pair-game/index.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        this.rightAnswers = 0;
+        this.openCards = [];
+    }
+
+    startGame() {
+        this.gameContainer.innerHTML = this.getCards();
+        this.addListeners();
+    }
+
+    addListeners() {
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.js:13]: Use [event delegation](https://javascript.info/event-delegation), please. It'll help you to increase the performance of your app.

File: submissions/ustymchyk/memory-pair-game/index.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        if (this.openCards.length < 2) {
+            this.openCards.push(btn);
+        }
+        if (this.openCards.length === 2) {
+            this.checkMatch();
+        }
+    }
+
+    checkMatch() {
+        if (this.openCards[0].getAttribute('data-id') === this.openCards[1].getAttribute('data-id')) {
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.js:34]: you can use this.openCards[0].dataset.id here as well
Also, maybe using ternary operator will help to make this code shorter:
```suggestion
        this.openCards[0].dataset.id === this.openCards[1].dataset.id ?  this.addMatch() : this.closeCards()
```

File: submissions/ustymchyk/memory-pair-game/index.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        }
+    }
+
+    addMatch() {
+        this.rightAnswers++;
+        this.openCards = [];
+
+        if (this.rightAnswers === 4) {
+            setTimeout(() => {
+                alert('Congratulations!');
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.js:47]: It's better to move this congrats string to the constants.

File: submissions/ustymchyk/memory-pair-game/index.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        } else {
+            this.closeCards();
+        }
+    }
+
+    addMatch() {
+        this.rightAnswers++;
+        this.openCards = [];
+
+        if (this.rightAnswers === 4) {
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.js:45]: Create a constant for `4` value. Now it looks like a magic number

File: submissions/ustymchyk/memory-pair-game/index.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            setTimeout(() => {
+                alert('Congratulations!');
+                this.startGame();
+            }, 400);
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.js:49]: It's better to move this value of `400` to the constants.

File: submissions/ustymchyk/memory-pair-game/index.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        }
+    }
+
+    closeCards() {
+        const cardsToClose = [...this.openCards];
+
+        setTimeout(() => {
+            cardsToClose.forEach(card => card.classList.remove('active'));
+        }, 1000);
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.js:58]: It's better to move this value of `1000` to the constants.

File: submissions/ustymchyk/memory-pair-game/index.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        return this.getShuffledIds().map(id => {
+            return `
+                <li class="card">
+                    <button class="card__btn" data-id="${id}" data-onclick="openCard()">
+                        <span class="card__front"></span>
+                        <span class="card__back" style="--image: url(${this.getPokemonImageUrl(id)})"></span>
+                    </button>
+                </li>
+            `;
+        }).join('');
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.js:73]: 👍🏻

File: submissions/ustymchyk/memory-pair-game/index.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            const methodName = attr.split('(')[0];
+
+            button.addEventListener('click', event => this[methodName](event));
+        });
+    }
+
+    openCard({ currentTarget: btn }) {
+        btn.classList.add('active');
+
+        if (this.openCards.length < 2) {
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.js:25]: Create a constant for `2` value. Now it looks like a magic number

File: submissions/ustymchyk/memory-pair-game/index.css [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  box-sizing: border-box;
+}
+
+:root {
+  --grey: #263238;
+  --yellow: rgb(247, 202, 75);
+  --blue: rgb(64, 98, 163);
+}
+
+body {
```
Comment (MaxKalashnyk) [inline submissions/ustymchyk/memory-pair-game/index.css:11]: Now your game field has quite big height on my screen. I'd recommend you decreasing it and adjust to the screen height.
Screenshot 2020-12-02 at 12 29 01

**Review by MaxKalashnyk — CHANGES_REQUESTED** [review]
good job! Look at my comments below.
Special respect for Pokemons! 🥇

**Review by MaxKalashnyk — CHANGES_REQUESTED** [review]
it looks much better. please look at my minor suggestion and we will ready to merge this PR

**Comment by ustymchyk** [comment]
@MaxKalashnyk done

**Review by MaxKalashnyk — APPROVED** [review]
Good job! Approved!

**Pull request merged**